_posts/2019-08-01-my-own-summer-neural-network.md

```diff
@@ -0,0 +1,321 @@
+tags:
+- Neural Networks
+- Machine Learning
+- TensorFlow
+- AI
+---
+
+I must confess: I have absolutely no knowledge on machine learning. Of course I know the basic idea and the concepts, but if before of my experiment described in this blog post someone would have asked me to build a program that can detect if an image is a hotdog, I wouldn't have had any idea where to start. So machine learning was definitely something I wanted to dig into during my summer vacation. Don't worry, I did not spent my whole vacation on my laptop, and this experiment was actually pretty fast to carry out. I would say this took me about 15 hours or so.
+
+Autonomous cars, which run purely on visual and other sensory input, are fascinating. Hence I initially wanted to build a simple car or a robot which would run on track autonomously. I also encountered a thing called [Donkey Car](https://www.donkeycar.com/), which is sort of an open source alternative for [AWS Deep Racer](https://aws.amazon.com/deepracer/).
+
+But the physical world poses problems:
+
+1. I don't have lots of knowledge about electronics
+2. I'm not particularly crafty person
+3. Gathering of the components would take money, and most of all time, and I wanted to get started quickly
+4. I would need a track which would take at least a good half of our living room and my partner might have an opinion on that
+
+Then I remembered really entertaining and fascinating video series by **sentdex** called [Python Plays: Grand Theft Auto V](https://www.youtube.com/playlist?list=PLQVvvaa0QuDeETZEOy4VdocT7TOjfSA8a). If you haven't watched that, you definitely should take couple of hours and watch - it's super educational and fun.
+
+And **BANG** it hit me while sipping a cup of joe on my backyard: I'll do a really simple game in which player "drives" and tries to avoid obstacles and then try to use techniques described in the autonomous GTA video series to teach my laptop to play my game. So it's a simulation of my future endeavours!
+
+You might ask why I didn't just pick a simple game from the internet? It's because I wanted to keep the machine learning part as simple as possible, and thought that the easiest way to achieve this is to have full control of the game mechanics.
+
+## The Game - Endless Flighter
+
+I poured another cup of joe, and rolled up some [Parcel](https://parceljs.org/) and [Three.js](https://threejs.org/). After couple of hours I had [Endless Flighter](https://github.com/juhofriman/endless-flighter) all set up and running on my browser.
+
+![Endless Flighter Human](/img/endless-flighter-human.gif)
+
+Almost as simple as possible game in which you can steer left or right and you should avoid obstacles. I did not even implement collision detection to my game, because I realised that it is not actually needed and it only makes things more complex. This way my learning data collection and testing the model is much more straight forward.
```

Comment:
```suggestion
Almost as simple as possible game in which you can steer left or right and you should avoid obstacles. I did not even implement collision detection, because I realized that it is not actually needed and it only makes things more complex. This way my learning data collection and testing the model is much more straightforward.
```